libnvme/src/firmware.rs

```diff
@@ -0,0 +1,306 @@
+        };
+
+        let number_of_slots =
+            unsafe { (*identify.inner).id_frmw.fw_nslot() } as usize;
+        let mut firmware_slots = Vec::with_capacity(number_of_slots);
+        for slot in &logpage.fw_frs[..number_of_slots] {
+            // FIXME there's probably a better way to do this? Basically the
+            // strings are packed into an array without a nul byte unless the
+            // slot itself is empty so we can't depend on `CStr::from_ptr()`.
+            // Instead we have to take each slice of bytes and convert it into a
+            // CString checking if the first byte is nul along the way.
+            let bytes: Vec<u8> = slot.iter().map(|&x| x as u8).collect();
+            if bytes[0] == b'\0' {
+                firmware_slots.push(None);
+                continue;
+            }
+            let cstring = unsafe { CString::from_vec_unchecked(bytes) };
+            // NVMe Spec: "The firmware revision is indicated as an ASCII
+            // string."
+            let firmware = cstring.to_string_lossy().trim().to_string();
+            firmware_slots.push(Some(firmware));
+        }
+
+        Self {
+            active_slot,
+            next_active_slot,
+            slot1_is_read_only,
+            number_of_slots,
+            firmware_slots,
+        }
+    }
+
+    /// Get the firmware version for a particular slot.
+    ///
+    /// Note that the NVMe spec allows for slots 1 - 7.
+    pub fn get_slot<S>(&self, slot: S) -> Result<Option<&String>, NvmeSlotError>
+    where
+        S: TryInto<NvmeSlot>,
+        S::Error: Into<NvmeSlotError>,
+    {
+        let slot = slot.try_into().map_err(Into::into)?;
+        // We subtract 1 because our internal mapping is 0 indexed.
+        match self.firmware_slots.get(slot.0 as usize - 1) {
+            Some(slot) => Ok(slot.as_ref()),
+            None => Err(NvmeSlotError::DoesNotExisit(slot.0)),
+        }
+    }
+
+    /// Returns an iterator over the firmware slots.
+    ///
+    /// The iterator yields a `Some(String)` if the slot has a firmware version
+    /// commited otherwise it yeilds `None`.
+    pub fn slot_iter(&self) -> impl Iterator<Item = Option<&String>> {
+        self.firmware_slots.iter().map(|o| o.as_ref())
+    }
+}
+
+impl<'a> Controller<'a> {
+    /// Get the controller's firmware log page.
+    pub fn get_firmware_log_page(&self) -> Result<FirmwareLogPage, NvmeError> {
+        let LogPageInfo { size, req, .. } =
+            self.get_logpage(LogPageName::Firmware)?;
+        // TODO the size of nvme_fwslot_log_t should always match the size
+        // returned from get_logpage. We may want a custom error type here if
+        // that's not true. Although for now libnvme should error if we give it
+        // something too small.
+        let mut buf = MaybeUninit::<nvme_fwslot_log_t>::zeroed();
+        self.check_result(
+            unsafe {
+                nvme_log_req_set_output(req.inner, buf.as_mut_ptr() as _, size)
```

Comment:
Nit - in my continued attempt to purge `as`, can this be

```suggestion
                nvme_log_req_set_output(req.inner, buf.as_mut_ptr().cast(), size)
```

instead? (Same comment on other pointer `as` casts)

Reply:
Fixed in dff4996b